core/domain/story_services.py

```diff
@@ -222,45 +226,115 @@ def _save_story(committer_id, story, commit_message, change_list):
     exps_dict = (
         exp_fetchers.get_multiple_explorations_by_id(exp_ids, strict=False))
 
-    for node in story.story_contents.nodes:
-        if (node.exploration_id is not None) and (
-                node.exploration_id not in exps_dict):
-            raise utils.ValidationError(
-                'Expected story to only reference valid explorations, '
-                'but found an exploration with ID: %s (was it deleted?)' %
-                node.exploration_id)
-
-    if exps_dict:
+    exp_rights = (
+        rights_manager.get_multiple_exploration_rights_by_ids(exp_ids))
+
+    exp_rights_dict = {}
+
+    for rights in exp_rights:
+        if rights is not None:
+            exp_rights_dict[rights.id] = rights.status
+
+    for exp_id in exp_ids:
+        if exp_id not in exps_dict:
+            error_string = (
+                'Expected story to only reference valid explorations, but found'
+                ' a reference to an invalid exploration with ID: %s'
+                % exp_id)
+            if raise_error:
+                raise utils.ValidationError(error_string)
+            validation_error_messages.append(error_string)
+        else:
+            if (
+                    exp_rights_dict[exp_id] !=
```

Comment:
Can this lead to `Keyerror`? (Considering we aren't populating exp_rights_dict with exp_id if rights is None.)

Reply:
Since for all valid explorations that are not deleted, a rights object would exist, and the case for deleted/invalid explorations is considered above, there shouldn't be an issue, right?

Reply:
Yeah, considering we have a prod_validation which makes sure that we have rights object for all exploration. (Thanks for the explanation, I'll mark it resolved!)